io.openems.edge.io.iot2000/src/io/openems/edge/io/iot2000/IOT2000.java

```diff
@@ -0,0 +1,222 @@
+
+		this.deactivateAllWriteChannels();
+	}
+
+	@Deactivate
+	protected void deactivate() {
+		this.deactivateAllWriteChannels();
+		super.deactivate();
+	}
+
+	@Override
+	public void handleEvent(Event event) {
+		if (!this.isEnabled()) {
+			return;
+		}
+		switch (event.getTopic()) {
+		case EdgeEventConstants.TOPIC_CYCLE_BEFORE_PROCESS_IMAGE:
+			this.updateChannels();
+			break;
+
+		case EdgeEventConstants.TOPIC_CYCLE_EXECUTE_WRITE:
+			this.eventExecuteWrite();
+			break;
+		}
+	}
+
+	private void eventExecuteWrite() {
+		for (int i = 0; i < this.digitalOutputChannels.length; i++) {
+			BooleanWriteChannel channel = this.digitalOutputChannels[i];
+			Boolean readValue = channel.value().get();
+			Optional<Boolean> writeValue = channel.getNextWriteValueAndReset();
+
+			if (!writeValue.isPresent()) {
+				// no write value
+				continue;
+			}
+			if (Objects.equals(readValue, writeValue.get())) {
+				// read value = write value
+				continue;
+			}
+			try {
+				//this.setOutputValue(this.gpioOutputMap[i], writeValue.get());
+				this.ioUtil.setOutput(i, writeValue.get());
+			} catch (Exception e) {
+				this.logError(this.log,
+						"Could not write Value to " + channel.address().toString() + " - Reason: " + e.getMessage());
+				e.printStackTrace();
+			}
+
+		}
+
+	}
+
+	private void updateChannels() {
+
+		this.updateIoChannels(this.digitalInputChannels, this.gpioInputMap);
+		this.updateIoChannels(this.digitalOutputChannels, this.gpioOutputMap);
+
+	}
+
+	private void updateIoChannels(Channel<?>[] channels, int[] gpioMap) {
+		for (int i = 0; i < channels.length; i++) {
+			try {
+				boolean val = this.ioUtil.getIoValue(gpioMap[i]);
+
+				Optional<Boolean> inOpt = Optional.ofNullable(val);
+
+				if (!channels[i].value().asOptional().equals(inOpt)) {
+
+					channels[i].setNextValue(val);
+				}
+
+			} catch (IOException e) {
+				this.logError(this.log, "Could not read Value for channel " + channels[i].address().toString()
+						+ " - Reason: " + e.getMessage());
+			}
+		}
+	}
+
+	@Override
+	public String debugLog() {
+		StringBuilder b = new StringBuilder();
+		b.append("IN: ");
+		for (BooleanReadChannel channel : this.digitalInputChannels) {
+			Optional<Boolean> valueOpt = channel.value().asOptional();
+			b.append(channel.channelId().name() + " ");
+			if (valueOpt.isPresent()) {
+				if (valueOpt.get()) {
+					b.append("1");
+				} else {
+					b.append("0");
+				}
+			} else {
+				b.append("-");
+			}
+			b.append(";");
+
+		}
+
+		b.append("  OUT: ");
+		for (BooleanWriteChannel channel : this.digitalOutputChannels) {
+			Optional<Boolean> valueOpt = channel.value().asOptional();
+			b.append(channel.channelId().name() + " ");
+			if (valueOpt.isPresent()) {
+				if (valueOpt.get()) {
+					b.append("1");
+				} else {
+					b.append("0");
+				}
+			} else {
+				b.append("-");
+			}
+			b.append(";");
+
+		}
+		return b.toString();
+	}
+
+	@Override
+	public BooleanReadChannel[] digitalInputChannels() {
```

Comment:
Maybe "getDigitalInputs()" to clarify that it is a getter function

Reply:
Cannot influence this as it is the implementation of the IO interface provided by OpenEMS.

Reply:
Maybe we should change it there ;)